File: src/main/java/eu/mjindra/attacks/Damage.java
package eu.mjindra.attacks;

import eu.mjindra.utils.dices.D0;
import eu.mjindra.utils.dices.Dice;

/**
 * Class representing damage.
 *
 * @author Martin Jindra
 * @version 30.05.2023
 */
public class Damage {

    public Dice dice;

    private DamageType type;

    public Damage(Dice dice, DamageType type) {
        this.dice = dice;
        this.type = type;
    }

    public Damage(byte modifier, DamageType type) {
        this.dice = new D0((byte) 1, modifier);
        this.type = type;
    }

    public Dice getDice() {
        return dice;
    }

    public void setDice(Dice dice) {
        this.dice = dice;
    }

    public DamageType getType() {
        return type;
    }

    public void setType(DamageType type) {
        this.type = type;
    }

    /**
     * Parses an expression for a Damage object.
     * A die (and optional damage type) is required.
     * @param expression expression
     * @return processed Damage object
     */
    public static Damage parse(String expression) {
        String[] stringSplit = expression.split(" ");

        if (stringSplit.length == 1) {
            stringSplit[0] = stringSplit[0].replace("+", "");
            return new Damage(Byte.parseByte(stringSplit[0]), DamageType.NONE);
        }
        DamageType type = DamageType.valueOf(stringSplit[1].toUpperCase());
        Dice damage = Dice.parse(stringSplit[0]);

        return new Damage(damage, type);
    }

    public String toString() {
        return String.format("%s %s", this.dice, this.type.desc());
    }
}

File: lib/dnd/combat/damage.dart
import 'package:helios/dnd/dices/d0.dart' show D0;
import 'package:helios/dnd/dices/die.dart';
import 'package:helios/dnd/dices/mixer.dart' show Mixer;
import 'package:helios/dnd/properties/damagetype.dart' show DamageType;

/// Class representing damage.
class Damage {
  Mixer dice = Mixer();
  DamageType type;

  Damage(this.dice, this.type);

  Damage.die(Die die, this.type) {
    dice.add(die);
  }

  Damage.mod(int modifier, this.type) {
    dice.add(D0.mod(modifier));
  }

  /// Parses an expression for a Damage object.
  /// A die (and optional damage type) is required.
  static Damage parse(String expression) {
    List<String> stringSplit = expression.trim().split(" ");

    if (stringSplit.length == 1) {
      stringSplit[0] = stringSplit[0].replaceAll("+", "");
      return Damage.mod(int.parse(stringSplit[0]), DamageType.none);
    }
    DamageType type;
    if (stringSplit[1] == '') {
      type = DamageType.none;
    } else {
      type = DamageType.values
          .firstWhere((DamageType e) => (e.name == stringSplit[1]));
    }
    return Damage(Mixer.parse(stringSplit[0]), type);
  }

  @override
  String toString() {
    return '$dice ${type == DamageType.none ? '' : type.name}';
  }
}
